### PR TITLE
feat: migrate to tf-sdk2

### PR DESCRIPTION
**Pull request by mavogel**
## Steps
see https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html
- [x] pass `ctx` to all methods
- [x] fix linters 
- [x] fix unit tests
- [x] fix builds. Need to update docker dependencies. See #119 
- [x] fix linters after update and rebasing
- [x] fix acceptance tests
- [x] replace ValidateFunc with `ValidateDiagFunc` 
- ~continue with the upgrade guide and add new [best practices for sdk v2](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html)~ -> will be done in #94 
- ~move website to docs directory. uses `index.md`~ -> will be done in #94 

## Notes
- we will not update to `ValidateDiagFunc` until https://github.com/hashicorp/terraform-provider-scaffolding/issues/47 is solved -> got solved. Will add it

### CODE REVIEW

**Comment by innovate-invent** [comment]
Have a look at some of the changes in #23. You will need to actually use the new context param for it to be effective.

**Comment by mavogel** [comment]
Thanks for pointing out! I made the rough changes yesterday and now I'll go for the details

File: internal/provider/resource_docker_container_funcs.go [hunk 1]
```diff
@@ -653,7 +653,16 @@ func resourceDockerContainerRead(ctx context.Context, d *schema.ResourceData, me
...
+	// TODO mavogel: move all fatteners to structures_container.go
+	extraHosts := make([]interface{}, len(container.HostConfig.ExtraHosts))
+	for i, extraHost := range container.HostConfig.ExtraHosts {
+		extraHostSplit := strings.Split(extraHost, ":")
+		extraHosts[i] = map[string]interface{}{
+			"host": extraHostSplit[0],
+			"ip":   extraHostSplit[1],
+		}
+	}
+	d.Set("host", extraHosts)
```
Comment (suzuki-shunsuke) [inline internal/provider/resource_docker_container_funcs.go:665]: Is this change required to migrate SDK?

File: internal/provider/resource_docker_service.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ func resourceDockerService() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"auth": {
-				Type:     schema.TypeMap,
+				Type:     schema.TypeList,
```
Comment (suzuki-shunsuke) [inline internal/provider/resource_docker_service.go:26]: State Migration should be implemented.

File: internal/provider/data_source_docker_registry_image_test.go [hunk 3]
```diff
@@ -46,6 +48,7 @@ func TestAccDockerRegistryImage_private(t *testing.T) {
 func TestAccDockerRegistryImage_auth(t *testing.T) {
 	registry := "127.0.0.1:15000"
 	image := "127.0.0.1:15000/tftest-service:v1"
+	ctx := context.TODO()
```
Comment (suzuki-shunsuke) [inline internal/provider/data_source_docker_registry_image_test.go:51]: Maybe context.Background is better?

File: internal/provider/provider.go [hunk 4]
```diff
@@ -12,152 +12,155 @@ import (
...
-		ResourcesMap: map[string]*schema.Resource{
-			"docker_container":      resourceDockerContainer(),
-			"docker_image":          resourceDockerImage(),
-			"docker_registry_image": resourceDockerRegistryImage(),
-			"docker_network":        resourceDockerNetwork(),
-			"docker_volume":         resourceDockerVolume(),
-			"docker_config":         resourceDockerConfig(),
-			"docker_secret":         resourceDockerSecret(),
-			"docker_service":        resourceDockerService(),
-			"docker_plugin":         resourceDockerPlugin(),
```
Comment (suzuki-shunsuke) [inline internal/provider/provider.go]: `docker_plugin` should not be removed.

**Review by suzuki-shunsuke — COMMENTED** [review]
Is it required to move `docker` package to `internal/provider` for SDK migration?

File: internal/provider/provider.go [hunk 5]
```diff
@@ -171,12 +171,12 @@ type AuthConfigs struct {
 }
 
 // Take the given registry_auth schemas and return a map of registry auth configurations
-func providerSetToRegistryAuth(authSet *schema.Set) (*AuthConfigs, error) {
+func providerSetToRegistryAuth(authList interface{}) (*AuthConfigs, error) {
```
Comment (suzuki-shunsuke) [inline internal/provider/provider.go:192]: Maybe `[]interface{}` is better.

File: internal/provider/resource_docker_service_funcs.go [hunk 6]
```diff
@@ -1044,7 +1044,9 @@ func createGenericResources(value interface{}) ([]swarm.GenericResource, error)
 // createRestartPolicy creates the restart poliyc of the service
 func createRestartPolicy(v interface{}) (*swarm.RestartPolicy, error) {
 	restartPolicy := swarm.RestartPolicy{}
-	rawRestartPolicy := v.(map[string]interface{})
+	rawRestartPolicySingleItem := v.([]interface{})
+	// because it's a list with MaxItems=1
+	rawRestartPolicy := rawRestartPolicySingleItem[0].(map[string]interface{})
```
Comment (suzuki-shunsuke) [inline internal/provider/resource_docker_service_funcs.go:1052]: Maybe the check of the length is required.

Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_service_funcs.go:1052]: Solved by a3fff814a0d079c5c355684ae6429264704ad4cc

File: internal/provider/provider.go (hunk 4 above)
Comment (suzuki-shunsuke) [reply internal/provider/provider.go]: Solved by 62714f67451c6f5fc1257ae6fd53d4276787f5d3

File: internal/provider/data_source_docker_plugin.go [hunk 7]
```diff
@@ -17,12 +17,15 @@ func dataSourceDockerPlugin() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
```
Comment (suzuki-shunsuke) [inline internal/provider/data_source_docker_plugin.go:23]: Ideally, this should be the other PR.

File: internal/provider/validators_test.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+import (
+	"testing"
+
+	"github.com/hashicorp/go-cty/cty"
+)
+
+func TestValidateIntegerGeqThan0(t *testing.T) {
+	v := 1
+	if diags := validateIntegerGeqThan(0)(v, *new(cty.Path)); len(diags) != 0 {
```
Comment (suzuki-shunsuke) [inline internal/provider/validators_test.go:11]:
```suggestion
	if diags := validateIntegerGeqThan(0)(v, *new(cty.Path)); diags.HasError() {
```

File: internal/provider/resource_docker_service.go (hunk 2 above)
Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_service.go:26]: https://www.terraform.io/docs/extend/resources/state-migration.html#terraform-v0-12-sdk-state-migrations

**Comment by suzuki-shunsuke** [comment]
Great work!

IMHO, we should do one thing in one pull request.
In case of this pull request, we should do only SDK migration in this pull request, and refactoring and bug fixes which are unrelated to the SDK migration should be done in the other pull requests.

I don't think we should split this pull request forcibly because it is hard,
but I wrote this comment for the future work.

File: internal/provider/resource_docker_container_funcs.go (hunk 1 above)
Comment (mavogel) [reply internal/provider/resource_docker_container_funcs.go:665]: Yes, this was necessary, because the tests failed after updating the sdk. and I wondered how they passed before.

File: internal/provider/resource_docker_service.go (hunk 2 above)
Comment (mavogel) [reply internal/provider/resource_docker_service.go:26]: Thanks for pointing out :) So for all `TypeMap` to `TypeList` we need those migrators right?

File: internal/provider/data_source_docker_registry_image_test.go (hunk 3 above)
Comment (mavogel) [reply internal/provider/data_source_docker_registry_image_test.go:51]: So basically in all `_test.go` files use `context.BackGround()` Can do this

File: internal/provider/provider.go (hunk 5 above)
Comment (mavogel) [reply internal/provider/provider.go:192]: I will take a look...

File: internal/provider/data_source_docker_plugin.go (hunk 7 above)
Comment (mavogel) [reply internal/provider/data_source_docker_plugin.go:23]: I know, but after updating the sdk, the tests failed because the attribute was missing. I guess the old sdk just covered this. This is why I added the property

**Comment by mavogel** [comment]
> Is it required to move docker package to internal/provider for SDK migration?

I wanted to be consistent with the latest [scaffolder](https://github.com/hashicorp/terraform-provider-scaffolding)

File: internal/provider/provider.go (hunk 5 above)
Comment (mavogel) [reply internal/provider/provider.go:192]: updated

File: internal/provider/resource_docker_service.go (hunk 2 above)
Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_service.go:26]: I think so.

File: internal/provider/data_source_docker_registry_image_test.go (hunk 3 above)
Comment (suzuki-shunsuke) [reply internal/provider/data_source_docker_registry_image_test.go]: I think so.
In my understanding, we should use `context.TODO()` when we plan to replace the context in future.
The purpose is to clarify as code that it should be replaced.

In case of test code, we don't have to replace the context so I think we shouldn't use `context.TODO()`.

File: internal/provider/data_source_docker_plugin.go (hunk 7 above)
Comment (suzuki-shunsuke) [reply internal/provider/data_source_docker_plugin.go:23]: I see. Thank you for your explanation.

File: internal/provider/data_source_docker_registry_image_test.go (hunk 3 above)
Comment (suzuki-shunsuke) [reply internal/provider/data_source_docker_registry_image_test.go]: Fixed by 9aca4fe

File: internal/provider/resource_docker_container_funcs.go (hunk 1 above)
Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_container_funcs.go:665]: I see. Thank you for your explanation.

File: internal/provider/validators_test.go (hunk 8 above)
Comment (suzuki-shunsuke) [reply internal/provider/validators_test.go]: Fixed by 27454eb

**Comment by suzuki-shunsuke** [comment]
> I wanted to be consistent with the latest [scaffolder](https://github.com/hashicorp/terraform-provider-scaffolding)

I got it. Looks Good.

File: internal/provider/resource_docker_service.go (hunk 2 above)
Comment (mavogel) [reply internal/provider/resource_docker_service.go]: - tested locally with unit tests and
- as described in the `contributing` section (I wonder how we can automate those tests)
```hcl
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      #version = "2.8.0"
      version = "9.9.9"
    }
  }
}

provider "docker" {
}

resource "docker_image" "foo" {
  name         = "nginx:latest"
  keep_locally = true
}

resource "docker_service" "foo" {
  name = "foo"
  task_spec {
    container_spec {
      image             = docker_image.foo.latest
      stop_grace_period = "10s"
    }
    # v2.8.0
    #restart_policy = { 
    # v9.9.9
    restart_policy {
      condition    = "on-failure"
      delay        = "3s"
      max_attempts = 4
      window       = "10s"
    }

  }

  endpoint_spec {
    ports {
      target_port    = "80"
      published_port = "8080"
    }
  }
}
```

File: internal/provider/resource_docker_service.go [hunk 9]
```diff
@@ -944,7 +944,7 @@ func resourceDockerServiceV0() *schema.Resource {
 		// callbacks are no longer relevant
 		Schema: map[string]*schema.Schema{
 			"auth": {
-				Type:     schema.TypeMap,
+				Type:     schema.TypeList,
```
Comment (suzuki-shunsuke) [inline internal/provider/resource_docker_service.go]: In my understanding, we should not change the old schema.

File: internal/provider/resource_docker_service.go (hunk 2 above)
Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_service.go]: Great work!
Would you split files per schema version like `docker_container` resource?

* resource_docker_service.go
* resource_docker_service_v1.go
* resource_docker_service_v0.go

I did the same in thing https://github.com/hashicorp/terraform-provider-docker/pull/272 .

File: internal/provider/resource_docker_service.go (hunk 9 above)
Comment (suzuki-shunsuke) [reply internal/provider/resource_docker_service.go]: Oh, it was solved.

Comment (mavogel) [reply internal/provider/resource_docker_service.go]: Yes, I reverted the change :)

**Comment by mavogel** [comment]
Feel free to leave a final review @suzuki-shunsuke :)

**Review by suzuki-shunsuke — APPROVED** [review]
🚀

**Comment by suzuki-shunsuke** [comment]
I can't merge this pull request.
I think we should fix the branch protection rule although I don't have the permission.

image
